### PR TITLE
[multimodal] Add Foundation Model for Object Detection

### PR DESCRIPTION
**Pull request by FANGAreNotGnu**
- Integrate [GroundingDino](https://github.com/IDEA-Research/GroundingDINO) into Autogluon to support open vocabulary detection.
- Add open vocabulary detection problem type.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by zhiqiangdon — COMMENTED** [review]
Per our offline discussion, we don't integrate the source code of grounding-dino into our repo. Instead, we let users install grounding-dino from source. Besides, we can submit two issues in the grounding-dino repo regarding the Pypi release and supporting tokenized tensor input for model.

File: .github/workflow_scripts/env_setup.sh [hunk 1]
```diff
@@ -53,6 +53,7 @@ function install_multimodal {
     mim install mmcv-full --timeout 60
     python3 -m pip install --upgrade "mmdet>=2.28, <3.0.0"
     python3 -m pip install --upgrade "mmocr<1.0"
+    python3 -m pip install --upgrade git+https://github.com/IDEA-Research/GroundingDINO.git
```
Comment (yinweisu) [inline .github/workflow_scripts/env_setup.sh:56]: Will this install the package from mainline? If so, this could lead to problem if the mainline is not stable (which is going to happen sometime more than likely). @Innixma @gradientsky Any opinions?

Comment (FANGAreNotGnu) [reply .github/workflow_scripts/env_setup.sh:56]: We only auto install it in CI, and ask users to install it manually when they try to use `open_vocabulary_detection`. I'll fork this repo and install from my github repo to make sure CI is not broken due to any unstable update, and will ask the author when it will be available on PYPI.

Comment (FANGAreNotGnu) [reply .github/workflow_scripts/env_setup.sh:56]: They just had a Pypi release. I'll test if it's working well and switch to it

Comment (FANGAreNotGnu) [reply .github/workflow_scripts/env_setup.sh:56]: The only pypi release does not seem like official. Just added an issue asking about official pypi release. But before the official release is available, shall we fork the repo or use the unofficial release?

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py [hunk 2]
```diff
@@ -485,6 +491,50 @@ def transform_rois(
...
+        for col_name in self._text_feature_names:
+            col_type = self._column_types[col_name]
+
+            if col_type == TEXT:
+                processed_data = df[col_name].tolist()
+            else:
+                raise ValueError(f"Unknown text type {col_type} for column {col_name}")
+
+            ret_data[col_name] = processed_data
+            ret_type[col_name] = self._column_types[col_name]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:534]: Can we just reuse `transform_text`?

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py [hunk 3]
```diff
@@ -254,6 +255,11 @@ def get_column_names(self, modality: str):
                 return self._image_path_names
             else:
                 return self._image_feature_names
+        elif modality == OVD:
+            if hasattr(self, "_image_path_names"):
+                return self._image_path_names + self._text_feature_names
+            else:
+                return self._image_feature_names + self._text_feature_names
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:262]: `ovd` is not a modality. I think no need to add this.

File: multimodal/src/autogluon/multimodal/data/process_ovd.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+from typing import Callable, Dict, List, Optional, Union
+
+import numpy as np
+import PIL
+from PIL import ImageFile
+from torch import nn
+
+from ..constants import COLUMN, IMAGE, IMAGE_META, IMAGE_PATH, IMAGE_VALID_NUM, PROMPT
+from .collator import ListCollator, PadCollator
+from .transform_bbox import *
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_ovd.py:12]: Which transforms from `transform_bbox` are used in this file?

File: multimodal/src/autogluon/multimodal/models/ovd.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        Returns
+        -------
+            A dictionary with logits and features.
+        """
+        try:
+            from groundingdino.util.utils import get_phrases_from_posmap
+        except ImportError as e:
+            logger.warning(
+                "Please install groundingdino by: pip install git+https://github.com/IDEA-Research/GroundingDINO.git"
+            )
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/ovd.py:117]: We already raise the error in `get_ovd_config_and_model`. Do we still need this warning?

File: multimodal/src/autogluon/multimodal/optimization/utils.py [hunk 6]
```diff
@@ -120,7 +121,7 @@ def get_loss_func(
             loss_func = nn.MSELoss()
     elif problem_type == NER:
         loss_func = nn.CrossEntropyLoss(ignore_index=0)
-    elif problem_type == OBJECT_DETECTION:
+    elif problem_type == OBJECT_DETECTION or problem_type == OPEN_VOCABULARY_OBJECT_DETECTION:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/optimization/utils.py:124]: `problem_type in [OBJECT_DETECTION, OPEN_VOCABULARY_OBJECT_DETECTION]` is better.

File: multimodal/src/autogluon/multimodal/presets.py [hunk 7]
```diff
@@ -292,6 +292,54 @@ def zero_shot_image_classification(presets: str = DEFAULT):
...
+        "model.names": ["ovd"],  # TODO: update with more preset hps
+        "env.num_gpus": 1,  # TODO: add multi gpu support and remove this
+        "env.eval_batch_size_ratio": 1,
+    }
+    hyperparameter_tune_kwargs = {}
+
+    if presets in [DEFAULT, BEST_QUALITY]:
+        hyperparameters.update(
+            {
+                "model.ovd.checkpoint_name": "GroundingDINO_SwinB",
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/presets.py:322]: Do they have larger backbones?

File: multimodal/src/autogluon/multimodal/utils/data.py [hunk 8]
```diff
@@ -277,6 +286,17 @@ def create_fusion_data_processors(
...
+            # create a multimodal processor for OVD.
+            data_processors[OVD].append(
+                create_data_processor(
+                    data_type=OVD,
+                    config=config,
+                    model=per_model,
+                )
+            )
+            if data_types is not None and IMAGE in data_types:
+                data_types.remove(IMAGE)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/data.py:299]: `OVD` is not a data type. If there are no correlations for processing image and text, we can create an image processor and a text processor for the ovd model.

File: multimodal/src/autogluon/multimodal/data/process_ovd.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                    "Finetuning/Evaluation with ground truth labels are not implemented for OVD yet"
+                )  # TODO
+            else:
+                ret[self.prompt_key] = per_col_content[0]
+        if self.requires_column_info:
+            pass  # TODO
+
+        ret.update(
+            {self.image_key: self.train_processor(image_data) if is_training else self.val_processor(image_data)}
+        )
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_ovd.py:146]: The image and text processing are independent. I think no need to put them in one processor.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 10]
```diff
@@ -2225,6 +2235,12 @@ def predict(
                     detection_classes=self._model.model.CLASSES,
                     result_path=None,
                 )
+            elif self._problem_type == OPEN_VOCABULARY_OBJECT_DETECTION:
+                pred = save_ovd_result_df(
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2239]: Why do we need to save results for ovd? `result_path=None` means not saving, but the function name is still misleading. By default, we need to return the same format as input.

File: .github/workflow_scripts/env_setup.sh [hunk 11]
```diff
@@ -53,6 +53,7 @@ function install_multimodal {
     mim install mmcv-full --timeout 60
     python3 -m pip install --upgrade "mmdet>=2.28, <3.0.0"
     python3 -m pip install --upgrade "mmocr<1.0"
+    python3 -m pip install --upgrade git+https://github.com/FANGAreNotGnu/GroundingDINO.git
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply .github/workflow_scripts/env_setup.sh]: Done, also used setup_mmcv

File: examples/automm/ovd/ovd_demo.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            "image": [sample_image_path],
+            "prompt": [
+                "Pink notice. Green sign. One Way sign. People group. Tower crane in construction. Lamp post. Glass skyscraper."
+            ],
+        },
+        as_pandas=True,
+    )
+    print("test_result:")
+    print(test_result)
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply examples/automm/ovd/ovd_demo.py]: Done

File: multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml [hunk 13]
```diff
@@ -150,6 +150,15 @@ model:
...
     output_bbox_format: "xyxy"
 
+  ovd:
+    checkpoint_name: "GroundingDINO_SwinB"
+    config_file: ""
+    data_types:
+      - "image"
+      - "text"
+    max_img_num_per_col: 1
+    output_bbox_format: "xyxy" # TODO
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/configs/model/fusion_mlp_image_text_tabular.yaml]: Added a link to explain different bbox formats

File: multimodal/src/autogluon/multimodal/configs/pretrain/ovd/grounding_dino/GroundingDINO_SwinB.cfg.py
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/configs/pretrain/ovd/grounding_dino/GroundingDINO_SwinB.cfg.py:1]: The name is from original grounding dino's repo. I'm trying to keep the config name the same as original ones.

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py (hunk 2 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:534]: It involves some work in text_processor. Will refactor it in next PR.

File: multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py (hunk 3 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/preprocess_dataframe.py:262]: Later will try to combine with Object Detection to use ROIS modality. But for now since the bounding box need special operation when we support training later, we might need to keep it.

File: multimodal/src/autogluon/multimodal/data/process_ovd.py (hunk 4 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/process_ovd.py]: I think this was only for previous implementation. Removed.

File: multimodal/src/autogluon/multimodal/models/ovd.py (hunk 5 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/models/ovd.py]: Removed.

File: multimodal/src/autogluon/multimodal/presets.py (hunk 7 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/presets.py:322]: They should have Swin-L but it is still not released.

File: multimodal/src/autogluon/multimodal/utils/data.py (hunk 8 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/data.py:299]: There will be  correlations for processing image and bounding boxes if we support training. Will refactor out text processor in next PR. And will keep OVD data type until we combine this and traditional detection (mmdet) to both use ROIS data type.

File: multimodal/src/autogluon/multimodal/data/process_ovd.py (hunk 9 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/process_ovd.py:145]: As above, will refactor out text processor in next PR. But have to keep image processing here for later training support.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 10 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:2241]: We are returning dict if as_pandas is False. Here we reuse save_ovd_result_df that both formatting result to df and save the df (if result_path is not None). Later we will add save result feature for ovd (using this function) together with evaluation for ovd.

File: multimodal/src/autogluon/multimodal/optimization/utils.py (hunk 6 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/optimization/utils.py]: Done

File: multimodal/src/autogluon/multimodal/utils/ovd.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    results = []
+
+    for image_pred, image_name in zip(pred, image_names):
+        results.append([image_name, image_pred])
+    result_df = pd.DataFrame(results, columns=["image", "bboxes"])
+
+    if result_path:
+        result_df.to_csv(result_path, index=False)
+        logger.info("Saved detection results to {}".format(result_path))
+        print("Saved detection results to {}".format(result_path))
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/ovd.py:40]: Need to remove all prints.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 10 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:2241]: `save_ovd_result_df ` seems to always return a panda dataframe, even though the input data is a dict.

File: multimodal/src/autogluon/multimodal/data/process_ovd.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+        Returns
+        -------
+        A dictionary containing one sample's images and their number.
+        """
+        ret = {}
+        image_data = {}
+
+        for per_col_name, per_col_content in texts_and_images.items():
+            if is_image_input(per_col_content, IMAGE_PATH):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/process_ovd.py:130]: We can use the `feature_modalities ` to check whether the modality type is image_path. Refer to `process_image.py`. `is_image_input ` is not necessary.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 10 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/predictor.py:2241]: Yes, that's also true for object detection. We only save panda dataframe. But will return list of dicts if input is not df.

File: multimodal/src/autogluon/multimodal/data/process_ovd.py (hunk 15 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/process_ovd.py]: Done.

File: multimodal/src/autogluon/multimodal/utils/ovd.py (hunk 14 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/ovd.py]: Done.

File: multimodal/src/autogluon/multimodal/data/utils.py [hunk 16]
```diff
@@ -421,6 +432,39 @@ def is_rois_input(sample):
     return isinstance(sample, list) and len(sample) and isinstance(sample[0], list) and len(sample[0]) == 5
 
 
+def is_image_input(images, image_type):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/data/utils.py:435]: We can remove this util since it's used only in one place.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 17]
```diff
@@ -110,7 +111,7 @@ def infer_metrics(
             eval_metric_name = SPEARMANR
         else:
             eval_metric_name = RMSE
-    elif problem_type == OBJECT_DETECTION:
+    elif problem_type == OBJECT_DETECTION or problem_type == OPEN_VOCABULARY_OBJECT_DETECTION:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:114]: `problem_type in [OBJECT_DETECTION, OPEN_VOCABULARY_OBJECT_DETECTION]` is better. Double check other places to see if such cases exist.

File: docs/tutorials/multimodal/object_detection/quick_start/quick_start_ovd.ipynb [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "aa00faab-252f-44c9-b8f7-57131aa8251c",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "!pip install autogluon.multimodal\n",
+    "#!pip install git+https://github.com/IDEA-Research/GroundingDINO.git  # uncomment this line if this package is not installed yet"
```
Comment (zhiqiangdon) [inline docs/tutorials/multimodal/object_detection/quick_start/quick_start_ovd.ipynb:31]: We can assume users already install automm? Instead, we should let users Install groundingdino?

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 17 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py]: Done.

File: docs/tutorials/multimodal/object_detection/quick_start/quick_start_ovd.ipynb (hunk 18 above)
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/quick_start/quick_start_ovd.ipynb]: Done. Will try import grounding dino first, if not installed, will import pip and use pip.main

File: docs/tutorials/multimodal/object_detection/quick_start/quick_start_ovd.ipynb [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+   "id": "8ce6e910",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pred = predictor.predict(\n",
+    "    {\n",
+    "        \"image\": [sample_image_path],\n",
+    "        \"prompt\": [\"Pink notice. Green sign. One Way sign. People group. Tower crane in construction. Lamp post. Glass skyscraper.\"],\n",
+    "    },\n",
+    "    as_pandas=True,\n",
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/quick_start/quick_start_ovd.ipynb:111]: Done

File: multimodal/src/autogluon/multimodal/data/utils.py (hunk 16 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/data/utils.py:435]: Reminder: consider removing this util and reverting the related changes.

Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/data/utils.py]: Done.

**Review by zhiqiangdon — APPROVED** [review]
LGTM. Considering adding a ovd unit test later.
